shell/src/main.rs

```diff
@@ -0,0 +1,162 @@
+                                "Switched to single line mode"
+                            };
+                            writeln!(stdout, "{mode_message}")?;
+                        } 
+                        else if line == ".quit" {
+                            writeln!(stdout, "Exiting shell...")?;
+                            break;
+                        }
+                        else if line == ".list" {
+                            writeln!(stdout, "Edited lines:")?;
+                            for line in buffer.clone().iter(){
+                                writeln!(stdout, "{line}")?;
+                            }
+                        }
+                        else if line == ".delete" || line == ".del" {
+                            if !buffer.is_empty() {
+                                let removed = buffer.pop().unwrap();
+                                writeln!(stdout, "Removed last line: {removed}")?;
+                            } else {
+                                writeln!(stdout, "Buffer is empty, nothing to delete")?;
+                            }
+                        }
+                        else if line == ".reset" {
+                            buffer.clear();
+                            rl.update_prompt(">>> ").expect("Can't update prompt");
+                            writeln!(stdout, "Buffer reset")?;
+                        }
+                        else if line == ".buffer" {
+                            writeln!(stdout, "Current buffer: {:?}", buffer)?;
+                        } 
+                        else {
+                            writeln!(stdout, "Unknown command: {line}")?;
+                        }
+                        continue;
+                    }
+                    
+                    rl.add_history_entry(line.clone());
+
+                    let command  = if (multiline){
+                        if buffer.is_empty() {
+                            if(line.is_empty()) {
+                                continue;
+                            }
+                            buffer = vec![line];
+                            rl.update_prompt("... ").expect("Can't update prompt");
+                            continue;
+                        }
+
+                        if(!line.is_empty()) {
+                            buffer.push(line);
+                            continue;
+                        }
+                        let command = buffer.join("\n");
+                        buffer.clear();
+                        rl.update_prompt(">>> ").expect("Can't update prompt");
+                        command
+                    } else {
+                        if line.is_empty() {
+                            continue;
+                        }
+                        line
+                    };
+
+                    writeln!(stdout, "Executing code: {command}")?;
+                    let result = interpreter.interpret(command).await;
```

Comment:
We cannot definitely await forever. If a program hangs there must be a way to kill the runtime graciously.